### ISSUE TITLE
orange-pi-zero 4.16.13 kernel cannot boot

### ISSUE THREAD

**Issue by nandra** [linux-sunxi/meta-sunxi#232]
When using sumo meta-sunxi build kernel (bzImage) just hangs after starting kernel. When build kernel with using sunxi_defconfig this on boots fine. Is this known issue? Thanks.

**Comment by nandra** [comment]
Looks like this simple change:
```
diff --git a/recipes-kernel/linux/linux.inc b/recipes-kernel/linux/linux.inc
index bef63e2..bf001a5 100644
--- a/recipes-kernel/linux/linux.inc
+++ b/recipes-kernel/linux/linux.inc
@@ -21,7 +21,7 @@ LOGO_SIZE ?= '${@oe.utils.conditional("MACHINE_GUI_CLASS", "bigscreen", "vga", "
 # accordingly.
 
 LOCALVERSION ?= ""
-KCONFIG_MODE ?= "alldefconfig"
+KCONFIG_MODE = "alldefconfig"
 KMACHINE ?= "${MACHINE}"
 
 #kernel_conf_variable CMDLINE "\"${CMDLINE} ${CMDLINE_DEBUG}\""
```

fix an issue

**Comment by nandra** [comment]
Looks like it's not fixing an issue. Looks like that sometimes zImage is build which have only 1,2M and this one is not able to boot

**Comment by geomatsi** [comment]
Hello @nandra 
What branches are you using for layers ?

Some time ago I have stumbled upon similar issue when using sumo branch of poky. Maybe rocko/pyro as well, but I don't remember exactly now.

In my case build was always successful, but sometimes kernel image was incorrect. In fact, that image corresponded to some default arm config, but not to sunxi_defconfig. But each time I could get proper kernel image just rebuilding kernel after the first build. 

It turned out that in my case the root cause was a race in build dependencies: bison-native was not ready during kernel compilation. The original fix is 81e8a52e8e40e47 in poky master. The minimal fix for the race can be as follows: add the following snippet
```
do_kernel_configme[depends] += "virtual/${TARGET_PREFIX}binutils:do_populate_sysroot"
do_kernel_configme[depends] += "virtual/${TARGET_PREFIX}gcc:do_populate_sysroot"
do_kernel_configme[depends] += "bc-native:do_populate_sysroot bison-native:do_populate_sysroot"
```
into recipes-kernel/linux/linux-mainline_4.16.13.bb.

Could you please let me know if it helps.

Regards,
Sergey

**Comment by bentech** [comment]
Hi, i've been getting a similar issue. This fixed my problem.

Thanks

**Comment by nandra** [comment]
@geomatsi using sumo branch when building image for opi-zero. Based on @bentech reply loks like this fix an issue. Can you please create PR (and would be possible to have it in sumo branch also)? Or should I do it? Thanks.

**Comment by geomatsi** [comment]
Hi @nandra 

Ok, I will send PR with this minimal fix to recipes-kernel/linux/linux.inc after I double-check the build and hear from you about the test results.

P.S.
In fact, I still in doubt where is the right place for this fix. It doesn't look like this fix belongs to meta-sunxi. On the other hand, I don't think its backport will be accepted to poky/sumo or any earlier poky branch because of the following observations:
* flex/bison are needed to build Linux kernel since v4.16
* the fix 81e8a52e8e40e47 exists in poky/master, which supplies kernel recipes for v4.18
* this fix is not available in poky/sumo branch, which is ok since the latest kernel recipe on that branch is for v4.15

Regards,
Sergey

**Comment by geomatsi** [comment]
Hi @nandra 
Hmm... Good point. I didn't notice commit 20e4d309e12bf10e2351e0016 in poky/sumo. Anyway, the original fix  81e8a52e8e40e47 on poky/master provides a thorough explanation:

_For linux-yocto, we also need to specify the compiler/tools dependencies for the configme task since it executes before configure and hence the main kernel build DEPENDS will not always be in the sysroot before it executes. Without those dependencies the kernel will be incorrectly configured (i.e. bison is missing) or the configuration will fail the mitigation tests._

Regards,
Sergey

**Comment by ebutera** [comment]
i merged #234, is this issue fixed now?

**Comment by nandra** [comment]
Yes issue is fixed. Thanks.
